test/mitmproxy/test_server.py

```diff
@@ -999,3 +999,50 @@ def handler(f):
         # (both terminated)
         # nothing happened here
         assert self.chain[1].tmaster.state.flow_count() == 2
+
+
+class AddServerCertsToClientChainMixin:
+
+    def test_add_server_certs_to_client_chain(self):
+        with open(self.servercert, "rb") as f:
+            d = f.read()
+        c1 = SSLCert.from_pem(d)
+        p = self.pathoc()
+        server_cert_found_in_client_chain = False
+        for cert in p.server_certs:
+            if cert.digest('sha256') == c1.digest('sha256'):
+                server_cert_found_in_client_chain = True
+                break
+        assert(server_cert_found_in_client_chain == self.add_server_certs_to_client_chain)
+
+
+class TestHTTPSAddServerCertsToClientChainTrue(tservers.HTTPProxyTest, AddServerCertsToClientChainMixin):
+
+    """
+    If --add-server-certs-to-client-chain is True, then the client should receive the server's certificates
+    """
+    add_server_certs_to_client_chain = True
+    ssl = True
+    servercert = tutils.test_data.path("data/trusted-server.crt")
+    ssloptions = pathod.SSLOptions(
+        cn="trusted-cert",
+        certs=[
+            ("trusted-cert", servercert)
+        ]
+    )
+
+
+class TestHTTPSAddServerCertsToClientChainFalse(tservers.HTTPProxyTest, AddServerCertsToClientChainMixin):
+
+    """
+    If --add-server-certs-to-client-chain is False, then the client should not receive the server's certificates
+    """
+    add_server_certs_to_client_chain = False
+    ssl = True
+    servercert = tutils.test_data.path("data/trusted-server.crt")
+    ssloptions = pathod.SSLOptions(
+        cn="trusted-cert",
+        certs=[
+            ("trusted-cert", servercert)
+        ]
+    )
```

Comment:
Excellent. I think we can even move `ssl`, `servercert` and `ssloptions` into `AddServerCertsToClientChainMixin` ? :smiley: